### PR TITLE
Validator announce

### PR DESCRIPTION
**Pull request by tkporter**
* Moves to a newer (unreleased) version of fuels-rs that allows Bytes to be returned. Note this doesn't yet allow Bytes to be passed in as inputs, so as a temporary workaround there's a `Vec<u8>` version used. Also note that the sway-lib String type doesn't have strong support in fuels-rs -- fuels-rs doesn't allow heap types like `Bytes` to be composed within other types, which is exactly what the String type is. So instead of using the String type, we just use the Bytes type instead for the interface
* Storage types are not yet composable in Sway. A `StorageBytes` type exists, which would ordinarily be a good candidate for storing the storage locations, but because we need to use a `StorageMapVec` such that multiple validators can have multiple storage locations stored, we can't compose these two. As a workaround, storage locations are stored as a `StorableString`, which uses a 128 character str (i.e. `str[128]`) under the hood, which is composable with storage. When https://github.com/FuelLabs/sway/pull/4297 gets in, we should consider moving away from StorableString. This is tracked in https://github.com/hyperlane-xyz/fuel-contracts/issues/66
* There's another workaround to account for fuels-rs not allowing heap types to compose with one another -- for example, you can't have `Vec<Bytes>` returned yet in fuels-rs. So instead a function `get_announced_storage_location` is provided to allow the looking up of a storage location for a single validator at a time that returns `Bytes`. It lets you specify the index of the storage location you want, or to default to the latest one. Unfortunately, we also can't have this function return `Option<Bytes>` because this isn't yet supported by fuels-rs (as it composes a heap type with another type...)
* This borrows the bytes_extended changes from #16 to get access to EvmAddress
* This also adds `signature_to_compact`, which helps convert from the ethers-rs Signature type to the [EIP 2098](https://eips.ethereum.org/EIPS/eip-2098) compatible signature representation that Fuel uses
*

### CODE REVIEW

File: test-utils/src/lib.rs [hunk 1]
```diff
@@ -17,6 +17,28 @@ pub fn bits256_to_h256(b: Bits256) -> H256 {
     H256(b.0)
 }
 
+// Fuel uses compact serialization for signatures following EIP 2098
+// See https://eips.ethereum.org/EIPS/eip-2098
+// |    32 bytes   ||           32 bytes           |
+// [256-bit r value][1-bit v value][255-bit s value]
+pub fn signature_to_compact(signature: &Signature) -> [u8; 64] {
```
Comment (yorhodes) [inline test-utils/src/lib.rs:24]: can we use https://github.com/FuelLabs/fuel-vm/blob/master/fuel-crypto/src/signature.rs ?

File: contracts/bytes-extended/src/main.sw [hunk 2]
```diff
@@ -76,26 +102,23 @@ impl Bytes {
     /// The Bytes will be able to hold exactly `length` bytes without
     /// reallocating.
     pub fn with_length(length: u64) -> Self {
-        // TODO: remove the + 1 once fixed upstream in the VM.
-        // An extra byte is allocated due to a bug in the FuelVM that prevents logging the
-        // byte closest to the top of the heap. This can occur if this is
-        // is the first time heap allocation is occurring.
-        // See https://github.com/FuelLabs/fuel-vm/issues/282 and the pending fix
-        // https://github.com/FuelLabs/fuel-vm/pull/287
-        let mut _self = Bytes::with_capacity(length + 1);
```
Comment (yorhodes) [inline contracts/bytes-extended/src/main.sw:85]: woot

**Review by yorhodes — APPROVED** [review]
thanks for putting in the work on the signatures
looks great!

File: test-utils/src/lib.rs (hunk 1 above)
Comment (tkporter) [reply test-utils/src/lib.rs:24]: oh good find

Looks like they have a way to convert from secp256k1::RecoverableSignature https://github.com/FuelLabs/fuel-vm/blob/master/fuel-crypto/src/signature.rs#L146 but I can't figure out how to go from an ethers-rs Signature to a RecoverableSignature...

File: contracts/validator-announce/tests/harness.rs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    let instance = ValidatorAnnounce::new(id.clone(), wallet);
+
+    (instance, id.into())
+}
+
+async fn sign_and_announce(
+    validator_announce: &ValidatorAnnounce<WalletUnlocked>,
+    signer: &Signers,
+    storage_location: String,
+) -> Result<FuelCallResponse<()>> {
```
[the remark this answers is not in the document]
Comment (tkporter) [reply contracts/validator-announce/tests/harness.rs:88]: Actually I think we're good, iiuc the validator already signs the announcement using ethers-rs in the expected manner. and then actually announcing it is done manually outside the validator

File: contracts/validator-announce/src/storable_string.sw [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+use std::bytes::Bytes;
+
+use bytes_extended::*;
+
+pub const MAX_STORABLE_STRING_CHARS: u64 = 128;
+
+// As a workaround to account for storable types not yet being composable,
+// which precludes the use of StorageBytes, Bytes are instead represented
+// as a str[128] (which is a fixed length array of bytes) and a corresponding length.
```
[the remark this answers is not in the document]
Comment (tkporter) [reply contracts/validator-announce/src/storable_string.sw:11]: lmao ikr

**Pull request merged**